Some CMake improvements for add_subdirectory() support

rokups opened the pull request:
Fixes sdl build when dlopen() is available but dlerror() is not. Regardless of missing dlerror() application builds and works.
Added multiple cmake options for specifying build configuration. This exposes options to cmake GUI on windows.
Moved some libAtomic-specific things into CMakeLists.txt file of that lib.
Other misc fixes that make engine more add_subdirectory() friendly in general and for android builds.

Also added few new cmake options that allow to opt out of:
* ATOMIC_JAVASCRIPT
* ATOMIC_DOTNET
* ATOMIC_EDITOR

Lets see if this breaks anything (hopefully not).

Edit: builds failing. Any way i could get access to build logs so i can fix the issues?

JoshEngebretson commented:
@rokups The CI boxes are local and I don't have the port open to the Jenkins install (ensuring Jenkins is and stays secure isn't on my todo list), there must be a way to publish the build logs somewhere though.. in the meantime:

Linux:

```
Linking CXX executable AtomicTool
../Atomic/libAtomic.a(Image.cpp.o): In function `Atomic::Image::SaveDDS(Atomic::String const&) const':
Image.cpp:(.text+0x15515): undefined reference to `squish::GetStorageRequirements(int, int, int)'
Image.cpp:(.text+0x1555a): undefined reference to `squish::CompressImage(unsigned char const*, int, int, void*, int, float*)'
../Atomic/libAtomic.a(OGLVertexBuffer.cpp.o): In function `Atomic::VertexBuffer::SetData(void const*)':
OGLVertexBuffer.cpp:(.text+0x99): undefined reference to `__glewBufferData'
../Atomic/libAtomic.a(OGLVertexBuffer.cpp.o): In function `Atomic::VertexBuffer::SetDataRange(void const*, unsigned int, unsigned int, bool)':
OGLVertexBuffer.cpp:(.text+0x4a9): undefined reference to `__glewBufferData'
OGLVertexBuffer.cpp:(.text+0x7fa): undefined reference to `__glewBufferSubData'
../Atomic/libAtomic.a(OGLVertexBuffer.cpp.o): In function `Atomic::VertexBuffer::Release()':
OGLVertexBuffer.cpp:(.text+0xcfc): undefined reference to `__glewDeleteBuffers'
../Atomic/libAtomic.a(OGLVertexBuffer.cpp.o): In function `Atomic::VertexBuffer::Create()':
OGLVertexBuffer.cpp:(.text+0xdfa): undefined reference to `__glewBufferData'
OGLVertexBuffer.cpp:(.text+0xf23): undefined reference to `__glewGenBuffers'
../Atomic/libAtomic.a(OGLIndexBuffer.cpp.o): In function `Atomic::IndexBuffer::SetData(void const*)':
OGLIndexBuffer.cpp:(.text+0x99): undefined reference to `__glewBufferData
```

Mac:

```
/Users/jenkins/Home/workspace/AtomicPR_MacEditor_iOS_Android/Artifacts/Build/Mac/Source/ThirdParty/Assimp/Release/libAssimp.a -Xlinker -dependency_info -Xlinker /Users/jenkins/Home/workspace/AtomicPR_MacEditor_iOS_Android/Artifacts/Build/Mac/Source/AtomicTool/Atomic.build/Release/AtomicTool.build/Objects-normal/x86_64/AtomicTool_dependency_info.dat -o /Users/jenkins/Home/workspace/AtomicPR_MacEditor_iOS_Android/Artifacts/Build/Mac/Source/AtomicTool/Release/AtomicTool
Undefined symbols for architecture x86_64:
  "_LZ4_compressBound", referenced from:
      ToolCore::ResourcePackager::WritePackageFile(Atomic::String const&) in libToolCore.a(ResourcePackager.o)
  "_LZ4_compressHC", referenced from:
      ToolCore::ResourcePackager::WritePackageFile(Atomic::String const&) in libToolCore.a(ResourcePackager.o)
  "pugi::xml_document::load(char const*, unsigned int)", referenced from:
      ToolCore::PlatformIOS::ParseProvisionAppIdentifierPrefix(Atomic::String const&) in libToolCore.a(PlatformIOS.o)
  "pugi::xml_document::xml_document()", referenced from:
      ToolCore::PlatformIOS::ParseProvisionAppIdentifierPrefix(Atomic::String const&) in libToolCore.a(PlatformIOS.o)
  "pugi::xml_document::~xml_document()", referenced from:
      ToolCore::PlatformIOS::ParseProvisionAppIdentifierPrefix(Atomic::String const&) in libToolCore.a(PlatformIOS.o)
  "Atomic::FileSystem::CreateDirs(Atomic::String const&, Atomic::String const&)", referenced from:
      ToolCore::NETCSProject::CreateCompileItemGroup(Atomic::XMLElement&) in libToolCore.a(NETProjectGen.o)
      ToolCore::NETCSProject::CreateMainPropertyGroup(Atomic::XMLElement&) in libToolCore.a(NETProjectGen.o)
      ToolCore::JSBind::GenerateCSharpBindings() in libToolCore.a(JSBind.o)
      ToolCore::JSBind::GenerateJavaScriptBindings() in libToolCore.a(JSBind.o)
      ToolCore::TextureImporter::Import() in libToolCore.a(TextureImporter.o)
  "Atomic::FileSystem::CreateDirsRecursive(Atomic::String const&)", referenced from:
      ToolCore::NETCSProject::CreateProjectFolder(Atomic::String const&) in libToolCore.a(NETProjectGen.o)
      ToolCore::NETSolution::Load(Atomic::JSONValue const&) in libToolCore.a(NETProjectGen.o)
      ToolCore::AndroidProjectGenerator::GenerateStringXML() in libToolCore.a(AndroidProjectGenerator.o)
      ToolCore::AtomicNETCopyAssemblies(Atomic::Context*, Atomic::String const&) in libToolCore.a(NETProjectSystem.o)
  "Atomic::FileSystem::SetLastModifiedTime(Atomic::String const&, unsigned int)", referenced from:
      ToolCore::AtomicNETCopyAssemblies(Atomic::Context*, Atomic::String const&) in libToolCore.a(NETProjectSystem.o)
````

JoshEngebretson commented:
Doing some CI work, need to get build errors reporting

@AtomicBuildBot retest this please

JoshEngebretson commented:
@AtomicBuildBot retest this please

rokups commented:
Just so noone freaks out seeing this mess im leaving a note on what is going. Scouting cmake 2.8.12 docs i noticed - surprise surprise - it has [target_compile_definitions](https://cmake.org/cmake/help/v2.8.12/cmake.html#command:target_compile_definitions), [target_compile_options](https://cmake.org/cmake/help/v2.8.12/cmake.html#command:target_compile_options) and [target_include_directories](https://cmake.org/cmake/help/v2.8.12/cmake.html#command:target_include_directories) along with `target_link_libraries` saying:

> Library dependencies are transitive by default. When this target is linked into another target then the libraries linked to this target will appear on the link line for the other target too.

Well damn, why noone used this in the past? I have no idea. But all this means there is no need for `_exported` macros and bunch of stuff in per-platform cmake scripts. So i am moving things that belong to the target into `CMakeLists.txt` of that target. Hopefully when im done build scripts will be way cleaner and simpler. While work is ongoing ill be abusing this PR and build system a lot. Please ignore bad commit messages and stupid commit content - when PR is done commits will be cleaned and squashed and final result will be perfectly clean PR.

JoshEngebretson commented:
@rokups Ok, thanks for the heads up, looking forward to seeing what you come up with... not needing the exported stuff from the last round would be great 👍

JoshEngebretson commented:
@rokups Congrats on first green build of CMake cleanup and refactor! ✔️ 🐫

rokups commented:
@JoshEngebretson please review these. Maybe it is ok to merge?

JoshEngebretson commented:
If ```ATOMIC_JAVASCRIPT``` or ```ATOMIC_DOTNET``` are turned on, script binding code needs to be generated, this is done by ToolCore (through the AtomicTool executable).  These targets should only build on desktop, and will not ever run on Android.

We need to make it possible to build Android solely using project C++ with JavaScript turned off, which will work with a single invocation of CMake.  For JS/C#, AtomicTool needs to be built for whatever the host desktop OS is and this can be handled in a number of ways.

JoshEngebretson commented on Build/CMake/Modules/AtomicCommon.cmake:
```diff
@@ -1,5 +1,5 @@
 #
-# Portions Copyright (c) 2008-2016 the Urho3D project.
+# Portions Copyright (c) 2008-2017 the Atomic project.
 #
```
Let's keep the Urho3D project copyright intact, some of the cmake script here is for compatibility, especially with ThirdParty

rokups replied:
In this specific case i changed copyright because `AtomicCommon.cmake` diverged from original file so much that syncing up would not be viable nor would it be a good idea since Urho's build system is upside-down and now we have that part fixed. So maybe this is ok after all? For this very specific reason i did not format any scripts in ThirdParty.

JoshEngebretson commented on CMakeLists.txt:
```diff
@@ -1,81 +1,64 @@
-
-if (NOT ATOMIC_BUILD_2D)
-    add_definitions_exported (-DATOMIC_PHYSICS -DATOMIC_3D)
-    add_link_libraries_exported (Bullet StanHull)
```
We lost the definition of ```ATOMIC_3D``` here, note ```ATOMIC_BUILD_2D``` used to be used for a purely 2D build, the motivation being to provide a small 2D web player.  We don't strictly need this right now, and seeing that we're building Recast/Detour/Crowd stuff on Web now.  The Web platform isn't currently hooked up to CI, other than when doing binary builds of editor.

```ATOMIC_3D``` does need to be defined, as there are generated guards for the 2D only builds in bindings.

rokups replied:
Yes indeed. Will fix. I am also renaming `ATOMIC_BUILD_2D` to `ATOMIC_2D_ONLY` as current option name is a bit vague and could be mistaken with enabling 2D support. If its not ok for some reason let me know and ill change it back ;)

Recast/detour should be working with emscripten though, right?

rokups commented:
The only script formatting i did was changing anything to 4 spaces. No tabs, never ever 🐫

rokups commented:
Changed license back in `AtomicCommon.cmake` because we must retain copyright after all. This brings me to the next thing i noticed - source files have Urho3D copyright removed. As per license it should be added back. It is perfectly fine to have multiple copyrights. I only added copyrights attributed to Atomic team to files that were either fully rewritten or newly created. It should appease license gods 🐫 

Also added `# ATOMIC BEGIN|END` around modifications in third party lib scripts and undone some `MSVC` mistakenly replaced with `WIN32`.

JoshEngebretson commented:
@rokups Urho3D copyright information should all be intact, if there is a file that is indeed under that copyright and it has been removed or modified incorrectly, that is an error that needs to be corrected.

JoshEngebretson commented:
Great work on the CMake improvements, I did a couple hours of testing and everything seems in order 👍   I squashed and will land from here: https://github.com/AtomicGameEngine/AtomicGameEngine/pull/1507

Thanks! 🚀 🐫 🚀
